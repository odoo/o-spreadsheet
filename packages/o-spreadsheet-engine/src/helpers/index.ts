export * from "./misc";
export * from "./numbers";
export * from "./references";
export * from "./uuid";
