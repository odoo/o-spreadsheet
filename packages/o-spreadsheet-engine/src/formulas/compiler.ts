import { parseNumber, unquote } from "../helpers";
import { _t } from "../translation";
import { CompiledFormula, FormulaToExecute } from "../types/base";
import { BadExpressionError, UnknownFunctionError } from "../types/errors";
import { DEFAULT_LOCALE } from "../types/locale";
import { FunctionCode, FunctionCodeBuilder, Scope } from "./code_builder";
import { parseTokens } from "./parser";
import { rangeTokenize } from "./range_tokenizer";
import { Token } from "./tokenizer";

interface ArgDefinitionLike {
  type?: string[];
}

interface FunctionDescriptionLike {
  args: ArgDefinitionLike[];
  minArgRequired: number;
  maxArgPossible: number;
  nbrArgRepeating: number;
  nbrArgOptional: number;
}

interface FunctionRegistryLike {
  content: Record<string, FunctionDescriptionLike>;
}

type ArgTargetingFn = (
  fn: FunctionDescriptionLike,
  argCount: number
) => (argIndex: number) => number | undefined;

let functionRegistryProvider: (() => FunctionRegistryLike) | null = null;
let argTargetingImpl: ArgTargetingFn | null = null;

export function setFunctionRegistryProvider(provider: () => FunctionRegistryLike) {
  functionRegistryProvider = provider;
}

export function setArgTargetingImplementation(fn: ArgTargetingFn) {
  argTargetingImpl = fn;
}

function getFunctionRegistry(): FunctionRegistryLike {
  if (!functionRegistryProvider) {
    throw new Error("Function registry provider has not been set.");
  }
  return functionRegistryProvider();
}

function getArgTargeting(): ArgTargetingFn {
  if (!argTargetingImpl) {
    throw new Error("argTargeting implementation has not been set.");
  }
  return argTargetingImpl;
}

export const OPERATOR_MAP = {
  "=": "EQ",
  "+": "ADD",
  "-": "MINUS",
  "*": "MULTIPLY",
  "/": "DIVIDE",
  ">=": "GTE",
  "<>": "NE",
  ">": "GT",
  "<=": "LTE",
  "<": "LT",
  "^": "POWER",
  "&": "CONCATENATE",
} as const;

export const UNARY_OPERATOR_MAP = {
  "-": "UMINUS",
  "+": "UPLUS",
  "%": "UNARY.PERCENT",
} as const;

interface LiteralValues {
  numbers: { value: number }[];
  strings: { value: string }[];
}

type InternalCompiledFormula = CompiledFormula & {
  literalValues: LiteralValues;
  symbols: string[];
};

export const functionCache: Record<string, FormulaToExecute> = {};

export function compile(formula: string): CompiledFormula {
  const tokens = rangeTokenize(formula);
  return compileTokens(tokens);
}

export function compileTokens(tokens: Token[]): CompiledFormula {
  try {
    return compileTokensOrThrow(tokens);
  } catch (error) {
    return {
      tokens,
      dependencies: [],
      execute() {
        return error;
      },
      isBadExpression: true,
      normalizedFormula: tokens.map((t) => t.value).join(""),
    };
  }
}

function compileTokensOrThrow(tokens: Token[]): CompiledFormula {
  const registry = getFunctionRegistry();
  const functions = registry.content;
  const argTargeting = getArgTargeting();

  const { dependencies, literalValues, symbols } = formulaArguments(tokens);
  const cacheKey = compilationCacheKey(tokens);
  if (!functionCache[cacheKey]) {
    const ast = parseTokens([...tokens]);
    const scope = new Scope();

    let stringCount = 0;
    let numberCount = 0;
    let dependencyCount = 0;

    if (ast.type === "BIN_OPERATION" && ast.value === ":") {
      throw new BadExpressionError(_t("Invalid formula"));
    }
    if (ast.type === "EMPTY") {
      throw new BadExpressionError(_t("Invalid formula"));
    }

    const compiledAST = compileAST(ast);
    const code = new FunctionCodeBuilder();
    code.append(`// ${cacheKey}`);
    code.append(compiledAST);
    code.append(`return ${compiledAST.returnExpression};`);
    const baseFunction = new Function(
      "deps",
      "ref",
      "range",
      "getSymbolValue",
      "ctx",
      code.toString()
    );

    // @ts-ignore - constructed function
    functionCache[cacheKey] = baseFunction;

    function compileFunctionArgs(astFuncall: any): FunctionCode[] {
      const { args } = astFuncall;
      const functionName = astFuncall.value.toUpperCase();
      const functionDefinition = functions[functionName];

      if (!functionDefinition) {
        throw new UnknownFunctionError(_t('Unknown function: "%s"', astFuncall.value));
      }

      assertEnoughArgs(args, functionDefinition, functionName);

      const compiledArgs: FunctionCode[] = [];
      const argToFocus = argTargeting(functionDefinition, args.length);

      for (let i = 0; i < args.length; i++) {
        const argIndex = argToFocus(i) ?? -1;
        const argDefinition =
          functionDefinition.args[argIndex] ??
          functionDefinition.args[i] ??
          ({} as ArgDefinitionLike);
        const currentArg = args[i];
        const argTypes = argDefinition?.type || [];
        const isMeta = argTypes.includes("META") || argTypes.includes("RANGE<META>");
        const hasRange = argTypes.some((t: string) => isRangeType(t));
        compiledArgs.push(compileAST(currentArg, isMeta, hasRange));
      }

      return compiledArgs;
    }

    function compileAST(ast: any, isMeta = false, hasRange = false): FunctionCode {
      const codeBuilder = new FunctionCodeBuilder(scope);
      if (ast.type !== "REFERENCE" && !(ast.type === "BIN_OPERATION" && ast.value === ":")) {
        if (isMeta) {
          throw new BadExpressionError(_t("Argument must be a reference to a cell or range."));
        }
      }
      if (ast.debug) {
        codeBuilder.append("debugger;");
        codeBuilder.append(`ctx["debug"] = true;`);
      }
      switch (ast.type) {
        case "BOOLEAN":
          return codeBuilder.return(`{ value: ${ast.value} }`);
        case "NUMBER":
          return codeBuilder.return(`this.literalValues.numbers[${numberCount++}]`);
        case "STRING":
          return codeBuilder.return(`this.literalValues.strings[${stringCount++}]`);
        case "REFERENCE":
          return codeBuilder.return(
            `${ast.value.includes(":") || hasRange ? "range" : "ref"}(deps[${dependencyCount++}], ${
              isMeta ? "true" : "false"
            })`
          );
        case "FUNCALL": {
          const args = compileFunctionArgs(ast).map((argCode) => argCode.assignResultToVariable());
          codeBuilder.append(...args);
          const fnName = ast.value.toUpperCase();
          return codeBuilder.return(`ctx['${fnName}'](${args.map((arg) => arg.returnExpression)})`);
        }
        case "UNARY_OPERATION": {
          const fnName = UNARY_OPERATOR_MAP[ast.value as keyof typeof UNARY_OPERATOR_MAP];
          const operand = compileAST(ast.operand, false, false).assignResultToVariable();
          codeBuilder.append(operand);
          return codeBuilder.return(`ctx['${fnName}'](${operand.returnExpression})`);
        }
        case "BIN_OPERATION": {
          const fnName = OPERATOR_MAP[ast.value as keyof typeof OPERATOR_MAP];
          const left = compileAST(ast.left, false, false).assignResultToVariable();
          const right = compileAST(ast.right, false, false).assignResultToVariable();
          codeBuilder.append(left);
          codeBuilder.append(right);
          return codeBuilder.return(
            `ctx['${fnName}'](${left.returnExpression}, ${right.returnExpression})`
          );
        }
        case "SYMBOL": {
          const symbolIndex = symbols.indexOf(ast.value);
          return codeBuilder.return(`getSymbolValue(this.symbols[${symbolIndex}])`);
        }
        case "EMPTY":
          return codeBuilder.return("undefined");
        default:
          return codeBuilder.return("undefined");
      }
    }
  }

  const compiledFormula: InternalCompiledFormula = {
    execute: functionCache[cacheKey],
    dependencies,
    literalValues,
    symbols,
    tokens,
    isBadExpression: false,
    normalizedFormula: cacheKey,
  };
  return compiledFormula;
}

function compilationCacheKey(tokens: Token[]): string {
  let cacheKey = "";
  for (const token of tokens) {
    switch (token.type) {
      case "STRING":
        cacheKey += "|S|";
        break;
      case "NUMBER":
        cacheKey += "|N|";
        break;
      case "REFERENCE":
      case "INVALID_REFERENCE":
        cacheKey += token.value.includes(":") ? "|R|" : "|C|";
        break;
      case "SPACE":
        break;
      default:
        cacheKey += token.value;
    }
  }
  return cacheKey;
}

function formulaArguments(tokens: Token[]) {
  const literalValues: LiteralValues = {
    numbers: [],
    strings: [],
  };
  const dependencies: string[] = [];
  const symbols: string[] = [];
  for (const token of tokens) {
    switch (token.type) {
      case "INVALID_REFERENCE":
      case "REFERENCE":
        dependencies.push(token.value);
        break;
      case "STRING":
        literalValues.strings.push({ value: unquote(token.value) });
        break;
      case "NUMBER":
        literalValues.numbers.push({ value: parseNumber(token.value, DEFAULT_LOCALE) });
        break;
      case "SYMBOL":
        symbols.push(unquote(token.value, "'"));
        break;
      default:
        break;
    }
  }
  return { dependencies, literalValues, symbols };
}

function assertEnoughArgs(
  args: any[],
  functionDefinition: FunctionDescriptionLike,
  functionName: string
) {
  const nbrArgSupplied = args.length;
  const { nbrArgRepeating, minArgRequired } = functionDefinition;

  if (nbrArgSupplied < minArgRequired) {
    throw new BadExpressionError(
      _t(
        "Invalid number of arguments for the %s function. Expected %s minimum, but got %s instead.",
        functionName,
        minArgRequired,
        nbrArgSupplied
      )
    );
  }

  if (nbrArgSupplied > functionDefinition.maxArgPossible) {
    throw new BadExpressionError(
      _t(
        "Invalid number of arguments for the %s function. Expected %s maximum, but got %s instead.",
        functionName,
        functionDefinition.maxArgPossible,
        nbrArgSupplied
      )
    );
  }

  if (nbrArgRepeating > 1) {
    const nbrValueRepeating =
      nbrArgRepeating * Math.floor((nbrArgSupplied - minArgRequired) / nbrArgRepeating);
    const nbrValueRemaining =
      nbrArgSupplied - minArgRequired - nbrValueRepeating - functionDefinition.nbrArgOptional;

    if (nbrValueRemaining > 0) {
      throw new BadExpressionError(
        _t(
          "Invalid number of arguments for the %s function. Repeatable arguments should be supplied in groups of %s, with up to %s optional. Got %s too many.",
          functionName,
          nbrArgRepeating,
          functionDefinition.nbrArgOptional,
          nbrValueRemaining
        )
      );
    }
  }
}

function isRangeType(type: string) {
  return type.startsWith("RANGE");
}
