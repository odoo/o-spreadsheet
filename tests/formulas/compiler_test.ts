import { functionCache } from "../../src/formulas/compiler";
import { compile, normalize } from "../../src/formulas/index";
import { functionRegistry } from "../../src/functions";
import { toZone } from "../../src/helpers";
import { CompiledFormula, NormalizedFormula, Range, ReturnFormatType } from "../../src/types";
import { evaluateCell } from "../helpers";

function compiledBaseFunction(formula: string): CompiledFormula {
  for (let f in functionCache) {
    delete functionCache[f];
  }
  compileFromCompleteFormula(formula);
  return Object.values(functionCache)[0];
}

function compileFromCompleteFormula(formula: string) {
  let formulaString: NormalizedFormula = normalize(formula);
  return compile(formulaString);
}

describe("expression compiler", () => {
  test.each(["=1", "=true", `="abc"`])("some arithmetic expressions", (formula) => {
    const compiledFormula = compiledBaseFunction(formula);
    expect(compiledFormula.toString()).toMatchSnapshot();
  });

  test("simple values that throw error", () => {
    expect(() => compiledBaseFunction(`='abc'`)).toThrowError();
  });

  test.each(["=1 + 3", "=2 * 3", "=2 - 3", "=2 / 3", "=-3", "=(3 + 1) * (-1 + 4)"])(
    "some arithmetic expressions",
    (formula) => {
      const compiledFormula = compiledBaseFunction(formula);
      expect(compiledFormula.toString()).toMatchSnapshot();
    }
  );

  test.each(["=sum(1,2)", '=sum(true, "")', "=sum(1,,2)"])(
    "some arithmetic expressions",
    (formula) => {
      const compiledFormula = compiledBaseFunction(formula);
      expect(compiledFormula.toString()).toMatchSnapshot();
    }
  );

  test("read some values and functions", () => {
    const compiledFormula = compiledBaseFunction("=A1 + sum(A2:C3)");
    expect(compiledFormula.toString()).toMatchSnapshot();
  });

  test("expression with $ref", () => {
    const compiledFormula = compiledBaseFunction("=$A1+$A$2+A$3");
    expect(compiledFormula.toString()).toMatchSnapshot();
  });

  test("expression with references with a sheet", () => {
    const compiledFormula = compiledBaseFunction("=Sheet34!B3");
    expect(compiledFormula.toString()).toMatchSnapshot();
  });

  test("expressions with a debugger", () => {
    const compiledFormula = compiledBaseFunction("=? A1 / 2");
    expect(compiledFormula.toString()).toMatchSnapshot();
  });

  test("async functions", () => {
    const compiledFormula = compiledBaseFunction("=WAIT(5)");
    expect(compiledFormula.toString()).toMatchSnapshot();
  });

  test("cells are converted to ranges if function require a range", () => {
    const compiledFormula = compiledBaseFunction("=sum(A1)");
    expect(compiledFormula.toString()).toMatchSnapshot();
  });

  test("cannot compile some invalid formulas", () => {
    expect(() => compiledBaseFunction("=qsdf")).toThrow();
  });
});

describe("compile dependencies format", () => {
  functionRegistry.add("ANYFUNCTION", {
    description: "any function",
    compute: () => 42,
    args: [{ name: "arg", description: "", type: ["ANY"], optional: true }],
    returns: ["NUMBER"],
  });

  const format = { specificFormat: "dd/mm/yy" };
  functionRegistry.add("RETURNFORMAT", {
    description: "a function with a specific return format",
    compute: () => 42,
    args: [
      { name: "arg1", description: "", type: ["ANY"], optional: true },
      { name: "arg2", description: "", type: ["ANY"], optional: true },
    ],
    returns: ["NUMBER"],
    returnFormat: format,
  });

  functionRegistry.add("RETURNARGSFORMAT", {
    description: "a function that returns a value in the format of the first argument",
    compute: () => 42,
    args: [
      { name: "arg1", description: "", type: ["ANY"], optional: true },
      { name: "arg2", description: "", type: ["ANY"], optional: true },
    ],
    returns: ["NUMBER"],
    returnFormat: ReturnFormatType.FormatFromArgument,
  });

  test.each(["=1", "=true", `="abc"`])("simple expressions don't return formats", (formula) => {
    const compiledFormula = compiledBaseFunction(formula);
    expect(compiledFormula.dependenciesFormat.length).toEqual(0);
  });

  test.each(["=A1", "=A1:B9", "=Sheet34!B3"])(
    "expression with ref return ref dependency",
    (formula) => {
      const compiledFormula = compiledBaseFunction(formula);
      expect(compiledFormula.dependenciesFormat).toEqual([0]);
    }
  );

  describe("expression with function", () => {
    test.each(["=ANYFUNCTION()", "=ANYFUNCTION(21)", "=ANYFUNCTION(TRUE)", "=ANYFUNCTION(B1)"])(
      "doesn't return formats, whatever args",
      (formula) => {
        const compiledFormula = compiledBaseFunction(formula);
        expect(compiledFormula.dependenciesFormat.length).toEqual(0);
      }
    );

    test.each(["=RETURNFORMAT()", "=RETURNFORMAT(21)", "=RETURNFORMAT(TRUE)", "=RETURNFORMAT(B1)"])(
      "that has a specific return format return this specific return format, whatever args",
      (formula) => {
        const compiledFormula = compiledBaseFunction(formula);
        expect(compiledFormula.dependenciesFormat).toEqual([format.specificFormat]);
      }
    );

    test.each([
      ["=RETURNARGSFORMAT()", []],
      ["=RETURNARGSFORMAT(21)", []],
      ["=RETURNARGSFORMAT(TRUE)", []],
      ["=RETURNARGSFORMAT(B1)", [0]],
      ["=RETURNARGSFORMAT(RETURNFORMAT(B1))", [format.specificFormat]],
      ["=RETURNARGSFORMAT(21,B1)", []],
    ])("that has 'ANY' as return format return the format of the first arg", (formula, result) => {
      const compiledFormula = compiledBaseFunction(formula);
      expect(compiledFormula.dependenciesFormat).toEqual(result);
    });
  });

  test.each([
    ["=+1", []],
    ["=+TRUE", []],
    ["=+B1", [0]],
    ["=+ANYFUNCTION(21)", []],
    ["=+RETURNFORMAT()", [format.specificFormat]],
    ["=+RETURNARGSFORMAT(B1)", [0]],
    ["=-1", []],
    ["=-TRUE", []],
    ["=-B1", [0]],
    ["=-ANYFUNCTION(21)", []],
    ["=-RETURNFORMAT()", [format.specificFormat]],
    ["=-RETURNARGSFORMAT(B1)", [0]],
  ])(
    "expression with unary operator returns dependency format of the right part",
    (formula, result) => {
      const compiledFormula = compiledBaseFunction(formula);
      expect(compiledFormula.dependenciesFormat).toEqual(result);
    }
  );

  describe("expression with bin operator that has 'ANY' as return format", () => {
    test.each([
      ["=1+B1", [0]],
      ["=B1+B2", [0, 1]],
      ["=B1+B2+B3", [0, 1, 2]],
      ["=B4+RETURNFORMAT()", [0, format.specificFormat]],
      ["=RETURNARGSFORMAT(B1,B2)+B3", [0, 2]],
      ["=ANYFUNCTION(B1)+B2", [1]],
      ["=RETURNARGSFORMAT(21,B1)+B2", [1]],
      ["=RETURNARGSFORMAT(B1+B2,B3)+B4", [0, 1, 3]],
      ["=RETURNARGSFORMAT(RETURNARGSFORMAT(B1,B2),B3)+B4", [0, 3]],
    ])("return dependencies format of the left and right part", (formula, result) => {
      const compiledFormula = compiledBaseFunction(formula);
      expect(compiledFormula.dependenciesFormat).toEqual(result);
    });

    test.each([
      ["=RETURNFORMAT()+B4", [format.specificFormat]],
      ["=B1+RETURNFORMAT()+B3", [0, format.specificFormat]],
      ["=RETURNARGSFORMAT(RETURNFORMAT(B1,B2),B3)+B4", [format.specificFormat]],
    ])(
      "return dependencies format of the left part if the left par is a string",
      (formula, result) => {
        const compiledFormula = compiledBaseFunction(formula);
        expect(compiledFormula.dependenciesFormat).toEqual(result);
      }
    );
  });
});

describe("compile functions", () => {
  describe("check number of arguments", () => {
    test("with basic arguments", () => {
      functionRegistry.add("ANYFUNCTION", {
        description: "any function",
        compute: () => {
          return true;
        },
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"] },
        ],
        returns: ["ANY"],
      });
      expect(() => compiledBaseFunction("=ANYFUNCTION()")).toThrow();
      expect(() => compiledBaseFunction("=ANYFUNCTION(1)")).toThrow();
      expect(() => compiledBaseFunction("=ANYFUNCTION(1,2)")).not.toThrow();
      expect(() => compiledBaseFunction("=ANYFUNCTION(1,2,3)")).toThrow();
    });

    test("with optional argument", () => {
      functionRegistry.add("OPTIONAL", {
        description: "function with optional argument",
        compute: (arg) => {
          return true;
        },
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"], optional: true },
        ],
        returns: ["ANY"],
      });
      expect(() => compiledBaseFunction("=OPTIONAL(1)")).not.toThrow();
      expect(() => compiledBaseFunction("=OPTIONAL(1,2)")).not.toThrow();
      expect(() => compiledBaseFunction("=OPTIONAL(1,2,3)")).toThrow();
    });

    test("with repeatable argument", () => {
      functionRegistry.add("REPEATABLE", {
        description: "function with repeatable argument",
        compute: (arg) => {
          return true;
        },
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"], optional: true, repeating: true },
        ],
        returns: ["ANY"],
      });
      expect(() => compiledBaseFunction("=REPEATABLE(1)")).not.toThrow();
      expect(() => compiledBaseFunction("=REPEATABLE(1,2)")).not.toThrow();
      expect(() => compiledBaseFunction("=REPEATABLE(1,2,3,4,5,6)")).not.toThrow();
    });

    test("with more than one repeatable argument", () => {
      functionRegistry.add("REPEATABLES", {
        description: "any function",
        compute: (arg) => {
          return true;
        },
        args: [
          { name: "arg1", description: "", type: ["ANY"] },
          { name: "arg2", description: "", type: ["ANY"], optional: true, repeating: true },
          { name: "arg3", description: "", type: ["ANY"], optional: true, repeating: true },
        ],
        returns: ["ANY"],
      });
      expect(() => compiledBaseFunction("=REPEATABLES(1, 2)")).toThrow();
      expect(() => compiledBaseFunction("=REPEATABLES(1, 2, 3)")).not.toThrow();
      expect(() => compiledBaseFunction("=REPEATABLES(1, 2, 3, 4)")).toThrow();
      expect(() => compiledBaseFunction("=REPEATABLES(1, 2, 3, 4, 5)")).not.toThrow();
    });
  });

  describe("with lazy arguments", () => {
    // this tests performs controls inside formula functions. For this reason, we
    // don't use mocked functions. Errors would be caught during evaluation of
    // formulas and not during the tests. So here we use a simple counter

    let count = 0;

    beforeAll(() => {
      functionRegistry.add("ANYFUNCTION", {
        description: "any function",
        compute: () => {
          count += 1;
          return true;
        },
        args: [],
        returns: ["ANY"],
      });

      functionRegistry.add("USELAZYARG", {
        description: "function with a lazy argument",
        compute: (arg) => {
          count *= 42;
          return arg();
        },
        args: [{ name: "lazyArg", description: "", type: ["ANY"], lazy: true }],
        returns: ["ANY"],
      });

      functionRegistry.add("NOTUSELAZYARG", {
        description: "any function",
        compute: (arg) => {
          count *= 42;
          return arg;
        },
        args: [{ name: "any", description: "", type: ["ANY"] }],
        returns: ["ANY"],
      });
    });

    test("with function as argument --> change the order in which functions are evaluated ", () => {
      count = 0;
      evaluateCell("A1", { A1: "=USELAZYARG(ANYFUNCTION())" });
      expect(count).toBe(1);
      count = 0;
      evaluateCell("A2", { A2: "=NOTUSELAZYARG(ANYFUNCTION())" });
      expect(count).toBe(42);
    });

    test.each(["=USELAZYARG(24)", "=USELAZYARG(1/0)"])(
      "functions call requesting lazy parameters",
      (formula) => {
        const compiledFormula = compiledBaseFunction(formula);
        expect(compiledFormula.toString()).toMatchSnapshot();
      }
    );
  });

  describe("with meta arguments", () => {
    beforeAll(() => {
      functionRegistry.add("USEMETAARG", {
        description: "function with a meta argument",
        compute: (arg) => arg,
        args: [{ name: "arg", description: "", type: ["META"] }],
        returns: ["STRING"],
      });
      functionRegistry.add("NOTUSEMETAARG", {
        description: "any function",
        compute: (arg) => arg,
        args: [{ name: "arg", description: "", type: ["ANY"] }],
        returns: ["ANY"],
      });
    });

    test.each(["=USEMETAARG(A1)", "=USEMETAARG(B2)"])(
      "function call requesting meta parameter",
      (formula) => {
        const compiledFormula = compiledBaseFunction(formula);
        expect(compiledFormula.toString()).toMatchSnapshot();
      }
    );

    test("throw error if parameter isn't cell/range reference", () => {
      expect(() => compiledBaseFunction("=USEMETAARG(X8)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG($X$8)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG(Sheet42!X8)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG('Sheet 42'!X8)")).not.toThrow();

      expect(() => compiledBaseFunction("=USEMETAARG(D3:Z9)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG($D$3:$Z$9)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG(Sheet42!$D$3:$Z$9)")).not.toThrow();
      expect(() => compiledBaseFunction("=USEMETAARG('Sheet 42'!D3:Z9)")).not.toThrow();

      expect(() => compiledBaseFunction('=USEMETAARG("kikou")')).toThrowError();
      expect(() => compiledBaseFunction('=USEMETAARG("")')).toThrowError();
      expect(() => compiledBaseFunction("=USEMETAARG(TRUE)")).toThrowError();
      expect(() => compiledBaseFunction("=USEMETAARG(SUM(1,2,3))")).toThrowError();
    });

    test("do not care about the value of the cell / range passed as a reference", () => {
      const compiledFormula1 = compileFromCompleteFormula("=USEMETAARG(A1)");
      const compiledFormula2 = compileFromCompleteFormula("=USEMETAARG(A1:B2)");
      const compiledFormula3 = compileFromCompleteFormula("=NOTUSEMETAARG(A1)");
      const compiledFormula4 = compileFromCompleteFormula("=NOTUSEMETAARG(A1:B2)");

      let refFn = jest.fn();
      let ensureRange = jest.fn();

      const ctx = { USEMETAARG: () => {}, NOTUSEMETAARG: () => {} };

      const rangeA1 = [{ zone: toZone("A1"), sheetId: "ABC" }] as Range[];
      const rangeA1ToB2 = [{ zone: toZone("A1:B2"), sheetId: "ABC" }] as Range[];

      compiledFormula1(rangeA1, "ABC", refFn, ensureRange, ctx);
      expect(refFn).toHaveBeenCalledWith(0, rangeA1, "ABC", true);
      expect(ensureRange).toHaveBeenCalledTimes(0);
      refFn.mockReset();

      compiledFormula2(rangeA1ToB2, "ABC", refFn, ensureRange, ctx);
      expect(refFn).toHaveBeenCalledWith(0, rangeA1ToB2, "ABC", true);
      expect(ensureRange).toHaveBeenCalledTimes(0);
      refFn.mockReset();

      compiledFormula3(rangeA1, "ABC", refFn, ensureRange, ctx);
      expect(refFn).toHaveBeenCalledWith(0, rangeA1, "ABC");
      expect(ensureRange).toHaveBeenCalledTimes(0);
      refFn.mockReset();

      compiledFormula4(rangeA1ToB2, "ABC", refFn, ensureRange, ctx);
      expect(refFn).toHaveBeenCalledWith(0, rangeA1ToB2, "ABC");
      expect(ensureRange).toHaveBeenCalledTimes(0);
      refFn.mockReset();
    });
  });
});
