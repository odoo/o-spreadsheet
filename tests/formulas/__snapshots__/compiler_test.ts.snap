// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compile functions with lazy arguments functions call requesting lazy parameters 1`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =USELAZYARG(24)
let _2 = ()=> 24
ctx.__lastFnCalled = 'USELAZYARG'
let _1 = ctx['USELAZYARG'](_2)
return _1;
}"
`;

exports[`compile functions with lazy arguments functions call requesting lazy parameters 2`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =USELAZYARG(1/0)
ctx.__lastFnCalled = 'DIVIDE'
let _2 = ()=> ctx['DIVIDE'](1, 0)
ctx.__lastFnCalled = 'USELAZYARG'
let _1 = ctx['USELAZYARG'](_2)
return _1;
}"
`;

exports[`compile functions with meta arguments function call requesting meta parameter 1`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =USEMETAARG(|0|)
let _2 = ref(0, deps, sheetId, true)
ctx.__lastFnCalled = 'USEMETAARG'
let _1 = ctx['USEMETAARG'](_2)
return _1;
}"
`;

exports[`compile functions with meta arguments function call requesting meta parameter 2`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =USEMETAARG(|0|)
let _2 = ref(0, deps, sheetId, true)
ctx.__lastFnCalled = 'USEMETAARG'
let _1 = ctx['USEMETAARG'](_2)
return _1;
}"
`;

exports[`expression compiler async functions 1`] = `
"async function anonymous(deps,sheetId,ref,range,ctx
) {
// =WAIT(5)
ctx.__lastFnCalled = 'WAIT'
let _1 = await ctx['WAIT'](5)
return _1;
}"
`;

exports[`expression compiler cells are converted to ranges if function require a range 1`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =sum(|0|)
let _2 = ref(0, deps, sheetId)
ctx.__lastFnCalled = 'SUM'
let _1 = ctx['SUM'](range(0, deps, sheetId))
return _1;
}"
`;

exports[`expression compiler expression with $ref 1`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =|0|+|1|+|2|
let _3 = ref(0, deps, sheetId)
let _4 = ref(1, deps, sheetId)
ctx.__lastFnCalled = 'ADD'
let _2 = ctx['ADD'](_3, _4)
let _5 = ref(2, deps, sheetId)
ctx.__lastFnCalled = 'ADD'
let _1 = ctx['ADD'](_2, _5)
return _1;
}"
`;

exports[`expression compiler expression with references with a sheet 1`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =|0|
let _1 = ref(0, deps, sheetId)
return _1;
}"
`;

exports[`expression compiler expressions with a debugger 1`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =? |0| / 2
debugger;
let _2 = ref(0, deps, sheetId)
ctx.__lastFnCalled = 'DIVIDE'
let _1 = ctx['DIVIDE'](_2, 2)
return _1;
}"
`;

exports[`expression compiler read some values and functions 1`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =|0| + sum(|1|)
let _2 = ref(0, deps, sheetId)
let _4 = ref(1, deps, sheetId)
ctx.__lastFnCalled = 'SUM'
let _3 = ctx['SUM'](range(1, deps, sheetId))
ctx.__lastFnCalled = 'ADD'
let _1 = ctx['ADD'](_2, _3)
return _1;
}"
`;

exports[`expression compiler some arithmetic expressions 1`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =1
return 1;
}"
`;

exports[`expression compiler some arithmetic expressions 2`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =true
return true;
}"
`;

exports[`expression compiler some arithmetic expressions 3`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =\\"abc\\"
return \\"abc\\";
}"
`;

exports[`expression compiler some arithmetic expressions 4`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =1 + 3
ctx.__lastFnCalled = 'ADD'
let _1 = ctx['ADD'](1, 3)
return _1;
}"
`;

exports[`expression compiler some arithmetic expressions 5`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =2 * 3
ctx.__lastFnCalled = 'MULTIPLY'
let _1 = ctx['MULTIPLY'](2, 3)
return _1;
}"
`;

exports[`expression compiler some arithmetic expressions 6`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =2 - 3
ctx.__lastFnCalled = 'MINUS'
let _1 = ctx['MINUS'](2, 3)
return _1;
}"
`;

exports[`expression compiler some arithmetic expressions 7`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =2 / 3
ctx.__lastFnCalled = 'DIVIDE'
let _1 = ctx['DIVIDE'](2, 3)
return _1;
}"
`;

exports[`expression compiler some arithmetic expressions 8`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =-3
ctx.__lastFnCalled = 'UMINUS'
let _1 = ctx['UMINUS']( 3)
return _1;
}"
`;

exports[`expression compiler some arithmetic expressions 9`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =(3 + 1) * (-1 + 4)
ctx.__lastFnCalled = 'ADD'
let _2 = ctx['ADD'](3, 1)
ctx.__lastFnCalled = 'UMINUS'
let _4 = ctx['UMINUS']( 1)
ctx.__lastFnCalled = 'ADD'
let _3 = ctx['ADD'](_4, 4)
ctx.__lastFnCalled = 'MULTIPLY'
let _1 = ctx['MULTIPLY'](_2, _3)
return _1;
}"
`;

exports[`expression compiler some arithmetic expressions 10`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =sum(1,2)
ctx.__lastFnCalled = 'SUM'
let _1 = ctx['SUM'](1,2)
return _1;
}"
`;

exports[`expression compiler some arithmetic expressions 11`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =sum(true, \\"\\")
ctx.__lastFnCalled = 'SUM'
let _1 = ctx['SUM'](true,\\"\\")
return _1;
}"
`;

exports[`expression compiler some arithmetic expressions 12`] = `
"function anonymous(deps,sheetId,ref,range,ctx
) {
// =sum(1,,2)
ctx.__lastFnCalled = 'SUM'
let _1 = ctx['SUM'](1,null,2)
return _1;
}"
`;
